src/cascadia/TerminalApp/TerminalWindow.cpp

```diff
@@ -0,0 +1,1262 @@
+// Arguments:
+// - warning: the SettingsLoadWarnings value to get the localized text for.
+// Return Value:
+// - localized text for the given warning
+static winrt::hstring _GetWarningText(SettingsLoadWarnings warning)
+{
+    return _GetMessageText(static_cast<uint32_t>(warning), settingsLoadWarningsLabels);
+}
+
+// Function Description:
+// - Creates a Run of text to display an error message. The text is yellow or
+//   red for dark/light theme, respectively.
+// Arguments:
+// - text: The text of the error message.
+// - resources: The application's resource loader.
+// Return Value:
+// - The fully styled text run.
+static Documents::Run _BuildErrorRun(const winrt::hstring& text, const ResourceDictionary& resources)
+{
+    Documents::Run textRun;
+    textRun.Text(text);
+
+    // Color the text red (light theme) or yellow (dark theme) based on the system theme
+    auto key = winrt::box_value(L"ErrorTextBrush");
+    if (resources.HasKey(key))
+    {
+        auto g = resources.Lookup(key);
+        auto brush = g.try_as<winrt::Windows::UI::Xaml::Media::Brush>();
+        textRun.Foreground(brush);
+    }
+
+    return textRun;
+}
+
+namespace winrt::TerminalApp::implementation
+{
+    TerminalWindow::TerminalWindow(const TerminalApp::SettingsLoadEventArgs& settingsLoadedResult) :
+        _settings{ settingsLoadedResult.NewSettings() },
+        _initialLoadResult{ settingsLoadedResult }
+    {
+        // The TerminalPage has to be constructed during our construction, to
+        // make sure that there's a terminal page for callers of
+        // SetTitleBarContent
+        _root = winrt::make_self<TerminalPage>();
+        _root->WindowProperties(*this);
+        _dialog = ContentDialog{};
+
+        // For your own sanity, it's better to do setup outside the ctor.
+        // If you do any setup in the ctor that ends up throwing an exception,
+        // then it might look like App just failed to activate, which will
+        // cause you to chase down the rabbit hole of "why is App not
+        // registered?" when it definitely is.
+    }
+
+    // Method Description:
+    // - Implements the IInitializeWithWindow interface from shobjidl_core.
+    HRESULT TerminalWindow::Initialize(HWND hwnd)
+    {
+        // Pass commandline args into the TerminalPage. If we were supposed to
+        // load from a persisted layout, do that instead.
+        auto foundLayout = false;
+        if (const auto& layout = LoadPersistedLayout())
+        {
+            if (layout.TabLayout().Size() > 0)
+            {
+                std::vector<Settings::Model::ActionAndArgs> actions;
+                for (const auto& a : layout.TabLayout())
+                {
+                    actions.emplace_back(a);
+                }
+                _root->SetStartupActions(actions);
+                foundLayout = true;
+            }
+        }
+        if (!foundLayout)
+        {
+            _root->SetStartupActions(_appArgs.GetStartupActions());

```

Comment:
why does TerminalWindow (per-window) have access to the commandline args (per-process)

Reply:
~Honestly, it _shouldn't_. I think somewhere in the history of these branches I tried moving it out of TerminalPage into AppLogic, but never could get it quite right when combined with state restoration.~


Oh wait I remember!

It's because it's _not_ per process. Commandline args _are_ per-window. Consider - you launch the Terminal, then run a `wt -w -1 -- foo`. That makes its own window. In this process, yes. But that new window has its own commandline args, separate from the ones that started the original process. 

Does that make sense?